### PR TITLE
[Relay] Esp32 Conversion & GPIO <-> API 

### PR DESCRIPTION
**Pull request by twoshark**
# Description

This pr converts the existing code to be compatible with esp32 wrooms. It also finishes the last mile to have the api control a gpio pin.

## ESP32 Conversion
- Eliminated the D1 mini build in platform io and replaced with espwroom32.
- Swapped 8266 libraries for esp32 compatible libraries.

## Multi-Core Architecture
- Divided the code to be run across 2 cores:
  - WebServer Task: runs wifimanager pages, handles all api requests and updates device state(s).
  - IO Task: Reads device state and updates output pin accordingly.
- Warning: Task code must now use `vTaskDelay(ms)` instead of `delay(ms)` to block only that task/core.

## Relay.h
- State & pin management class for the relay device.
- Provides GetState() and SetState() methods with locking so they maybe used across cores/tasks.
- Handler() is the only entity that updates the output pin.

## API
- Took steps towards genericizing the wifimanagerwrapper but it still needs some more work to be device generic. A DI pattern maybe, or just an inversion of dependencies between wfm & the devices.
- Maintained the REST style endpoints for now. We'll re-think that as a future effort to genericize the wfm wrapper.
- Connected endpoints to Relay class's Set method.

# Testing

- Manually verified that it produces a healthy set of wifimanager pages and does control the state in the Relay class
- Pending: verify with a connected relay

# CI

- Added a GH action to run `pio run` on every commit

### CODE REVIEW

File: lib/Relay/Relay.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            } else {
+                digitalWrite(this->pin, LOW);
+            }
+            this->unlockState();
+        }
+};
+
+
+
+#endif
```
Comment (twoshark) [inline lib/Relay/Relay.h]:
```suggestion
#endif

```

File: .vscode/settings.json [hunk 2]
```diff
@@ -8,6 +8,17 @@
...
+        "random": "cpp",
+        "fstream": "cpp",
+        "iosfwd": "cpp",
+        "istream": "cpp",
+        "ostream": "cpp",
+        "sstream": "cpp",
+        "streambuf": "cpp"
+    },
+    "C_Cpp.errorSquiggles": "enabled"
 }
```
Comment (twoshark) [inline .vscode/settings.json]:
```suggestion
}

```

File: platformio.ini [hunk 3]
```diff
@@ -6,14 +6,16 @@
...
+platform = espressif32
 framework = arduino
+board = upesy_wroom
+monitor_speed = 115200
 build_src_filter = +<relay/>
 lib_deps = 
-	tzapu/WiFiManager@^0.16.0
-	bblanchon/ArduinoJson@^6.21.3
+	WiFiManager@2.0.16-rc.2
+	bblanchon/ArduinoJson@^6.21.3
```
Comment (twoshark) [inline platformio.ini]:
```suggestion
	bblanchon/ArduinoJson@^6.21.3

```

File: .github/workflows/platformio.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+            ~/.platformio/.cache
+          key: ${{ runner.os }}-pio
+      - uses: actions/setup-python@v4
+        with:
+          python-version: '3.9'
+      - name: Install PlatformIO Core
+        run: pip install --upgrade platformio
+
+      - name: Build PlatformIO Project
+        run: pio run
```
Comment (twoshark) [inline .github/workflows/platformio.yaml]:
```suggestion
        run: pio run
```

Comment (twoshark) [reply .github/workflows/platformio.yaml]:
```suggestion
        run: pio run

```

File: .github/workflows/platformio.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+name: PlatformIO CI
```
Comment (dtp263) [inline .github/workflows/platformio.yaml:1]: love this

File: README.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+This repo contains all esp32 dc-compliant devices and shared libraries.
+
+## Building and Uploading
+
+[PlatformIO](https://platformio.org/) - We reccommend using PlatformIO to build and upload the devices in this repo. 
+
+## Reccommended IDE
+
+[VSCode](https://code.visualstudio.com/) - This repo include a `.vscode/` folder which contains common settings for PlatformIO and Cpp compilation. You are welcome to use another IDE, but may have to manually adjust your plugin settings.
```
Comment (dtp263) [inline README.md]:
```suggestion
[VSCode](https://code.visualstudio.com/) - This repo includes a `.vscode/` folder which contains common settings for PlatformIO and Cpp compilation. You are welcome to use another IDE, but may have to manually adjust your plugin settings. 
```

File: lib/WiFiManagerWrapper/WiFiManagerWrapper.h [hunk 7]
```diff
@@ -265,16 +272,16 @@ class WiFiManagerWrapper
...
-        ESP.eraseConfig(); 
-        delay(2000);
-        ESP.reset();
+        WiFi.disconnect(true);
+        vTaskDelay(2000);
+        ESP.restart();
       } else if (path == "/output/on") {
         Serial.println("Output on");
-        outputState = "on";
+        this->relay->SetState(OPEN);
```
Comment (dtp263) [inline lib/WiFiManagerWrapper/WiFiManagerWrapper.h:280]: This should somehow be outside of the WiFiManager wrapper. My thoughts are that we want any device to be able to use the WiFiWrapper. But also setup their own endpoints?

This could def be a larger change so totally cool if you think this should be done in a different PR.

Comment (dtp263) [reply lib/WiFiManagerWrapper/WiFiManagerWrapper.h:280]: Which I see you mentioned in your PR description... lol

Yeah we can figure something out here soon.

**Review by dtp263 — APPROVED** [review]
Sweeeetttt
